File: 435 Non-overlapping Intervals/435 Non-overlapping Intervals.go
func eraseOverlapIntervals(intervals [][]int) int {
    sort.Sort(byIntervalEnd(intervals))

    if len(intervals) == 0 {
        return 0
    }
    var remove int
    current := intervals[0]
    for i := 1; i < len(intervals); i++ {
        if intervals[i][0] < current[1] {
            remove++
        } else {
            current = intervals[i]
        }
    }
    return remove
}

type byIntervalEnd [][]int

func (b byIntervalEnd) Len() int {
    return len(b)
}

func (b byIntervalEnd) Swap(i, j int) {
    b[i], b[j] = b[j], b[i]
}

func (b byIntervalEnd) Less(i, j int) bool {
    return b[i][1] < b[j][1]
}

File: 435 Non-overlapping Intervals/435 Non-overlapping Intervals.py
"""
Given an array of intervals intervals where intervals[i] = [starti, endi], return the minimum number of intervals you need to remove to make the rest of the intervals non-overlapping.

Example 1:

Input: intervals = [[1,2],[2,3],[3,4],[1,3]]
Output: 1
Explanation: [1,3] can be removed and the rest of the intervals are non-overlapping.

Example 2:

Input: intervals = [[1,2],[1,2],[1,2]]
Output: 2
Explanation: You need to remove two [1,2] to make the rest of the intervals non-overlapping.

Example 3:

Input: intervals = [[1,2],[2,3]]
Output: 0
Explanation: You don't need to remove any of the intervals since they're already non-overlapping.
"""
class Solution:
    def eraseOverlapIntervals(self, intervals: List[List[int]]) -> int:
        intervals = sorted(intervals, key=lambda x: x[1])
        end = float('-inf')
        count = 0
        [[1,2],[1,3],[2,3],[3,4]]

        for interval in intervals:
            if interval[0] >= end:
                end = interval[1]
            else:
                count += 1

        return count
